phoenix-core/src/main/java/org/apache/phoenix/schema/transform/TransformMaintainer.java

```diff
@@ -0,0 +1,458 @@
+        boolean encodedColumnNames = maintainer.newTableEncodingScheme != NON_ENCODED_QUALIFIERS;
+        Iterator<ServerCachingProtos.ColumnReference> newTableColRefItr = newTableColRefsForCoveredColumnsList.iterator();
+        for (ServerCachingProtos.ColumnReference colRefFromProto : oldTableColRefsForCoveredColumnsList) {
+            ColumnReference oldTableColRef = new ColumnReference(colRefFromProto.getFamily().toByteArray(), colRefFromProto.getQualifier().toByteArray());
+            ColumnReference newTableColRef;
+            if (encodedColumnNames) {
+                ServerCachingProtos.ColumnReference fromProto = newTableColRefItr.next();
+                newTableColRef = new ColumnReference(fromProto.getFamily().toByteArray(), fromProto.getQualifier().toByteArray());
+            } else {
+                byte[] cq = oldTableColRef.getQualifier();
+                byte[] cf = oldTableColRef.getFamily();
+                newTableColRef = new ColumnReference(cf, cq);
+            }
+            maintainer.coveredColumnsMap.put(oldTableColRef, newTableColRef);
+        }
+        maintainer.logicalNewTableName = proto.getLogicalNewTableName();
+        maintainer.initCachedState();
+        return maintainer;
+    }
+
+
+    public static List<IndexMaintainer> deserialize(byte[] buf) {
+        return deserialize(buf, 0, buf.length);
+    }
+
+    private static List<IndexMaintainer> deserialize(byte[] buf, int offset, int length) {
+        List<IndexMaintainer> maintainers = Collections.emptyList();
+        if (length > 0) {
+            ByteArrayInputStream stream = new ByteArrayInputStream(buf, offset, length);
+            DataInput input = new DataInputStream(stream);
+            try {
+                int size = WritableUtils.readVInt(input);
+                boolean isDataTableSalted = size < 0;
+                size = Math.abs(size);
+                RowKeySchema rowKeySchema = new RowKeySchema();
+                rowKeySchema.readFields(input);
+                maintainers = Lists.newArrayListWithExpectedSize(size);
+                for (int i = 0; i < size; i++) {
+                    int protoSize = WritableUtils.readVInt(input);
+                    byte[] b = new byte[protoSize];
+                    input.readFully(b);
+                    ServerCachingProtos.TransformMaintainer proto = ServerCachingProtos.TransformMaintainer.parseFrom(b);
+                    maintainers.add(TransformMaintainer.fromProto(proto, rowKeySchema, isDataTableSalted));
+                }
+            } catch (IOException e) {
+                throw new RuntimeException(e); // Impossible
+            }
+        }
+        return maintainers;
+    }
+
+    // Return new table's name
+    public byte[] getIndexTableName() {
+        return newTableName;
+    }
+
+    public byte[] buildRowKey(ValueGetter valueGetter, ImmutableBytesWritable rowKeyPtr, byte[] regionStartKey, byte[] regionEndKey, long ts)  {
```

Comment:
nit: Javadoc would be really useful here. (Have to read the method to figure out that we're building a new table key from the old table key, rather than vice versa.)